Edit "What are Godot classes"

NathanLovato opened the pull request:
Changes:

- Simplify the article
- Change the title to better reflect the guide's point
- Try to make it clearer that scripts and scenes are not actually
- classes, but rather to be thought of as if they were classes in an OO
- language; OO principles apply to them
- Move section about the performances of scripts vs PackedScene to scenes_versus_scripts.rst

Resolves #3910

*Bugsquad edit:* Preview articles [here](https://github.com/NathanLovato/godot-docs/blob/edit/what-are-godot-classes/getting_started/workflow/best_practices/scenes_versus_scripts.rst) and [here](https://github.com/NathanLovato/godot-docs/blob/edit/what-are-godot-classes/getting_started/workflow/best_practices/what_are_godot_classes.rst).

NathanLovato commented:
Here's a first pass at the article. After shortly talking with @aaronfranke yesterday and re-reading both the current and original versions, I felt it needed a bit of a split and some simplification.

I tried to focus on one point: many best practices in Godot are about applying OO principles to scenes and scripts. This version strips some details and extra notes from the original. The part about performances of scenes vs scripts moved to the scene vs script article.

willnationsdev commented on getting_started/workflow/best_practices/scenes_versus_scripts.rst:
```diff
@@ -147,6 +147,63 @@ this way.
 
 On the downside, it also means having to use largely imperative programming.
 
+Performances of scripts vs PackedScenes
```
I've personally always thought that the plural form, `Performances`, just sounds *incredibly* awkward in English. You only use it when referring to specific events of evaluating them, not when describing their performance in general.

I'd just drop the 's' to make it more clean. And change `scripts` to `Script`. If that's not enough of a reason, then you could also add more detail to the topic: `Comparing the performance of Script vs PackedScene`.

willnationsdev reviewed (commented):
Nice! I like the way things are split up here. :-D

NathanLovato commented:
Thanks, Will, just addressed your comments.

Feel free to use the suggestion feature and, anyone, if you have merge rights, to use it to edit the PR directly. I'm not pretty about my code or text, you're welcome to modify it anytime.

willnationsdev commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -1,127 +1,66 @@
-        }
-    }
+    Even scripts that don't use the ``extends`` keyword implicitly inherit from the engine's base
+    :ref:`Reference <class_Reference>` class. This implicit type allows the object to defer to the
+    script's content where the engine logic deems appropriate.
```
`This implicit type allows` -> `The native type inheritance allows`

Reason: it isn't just the "implicit" type that enables this behavior, but just the existence of an underlying type at all.

aaronfranke replied:
Maybe just delete this note? Doesn't everything run by the engine that can have a script attached have the engine checking that script first? Also, isn't it that the script overrides the engine methods, not that the engine defers to them?

willnationsdev replied:
> Also, isn't it that the script overrides the engine methods, not that the engine defers to them?

This is actually true in the general case. I think what I had in mind while writing that statement initially were situations where the engine directly attempts to call a method on a script (like notification callbacks, declarative code methods, etc.). In those cases, the engine directly says, "does it have a script? Does the script have this method? If so, call it."

> Maybe just delete this note?

Yeah. Re-examining it, it doesn't actually add much to the content of the article since it effectively restates existing information in a different way. Seems so anyway.

willnationsdev commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -1,127 +1,66 @@
-Script code like this is much slower than engine-side C++ code. Each change
-makes a separate call to the scripting API which leads to many "look-ups" on the
-back-end to find the logic to execute.
+    As a result, you can instance scripts without the ``extends`` keyword from code, but you cannot
+    attach them to a :ref:`Node <class_Node>`.
```
```rst
As a result, you can instantiate scripts without the ``extends`` keyword from code.
Since they extend ``Reference`` though, you cannot attach them to a :ref:`Node <class_Node>`.
```

Reason: more clarity.

willnationsdev commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -130,10 +69,11 @@ The content of a scene helps to define:
 
-The scene is *always an extension of the script attached to its root node*. You
-can see all the nodes it contains as part of a single class.
+The scene is *always an extension of the script attached to its root node* so you can interpret it
```
`so you can` -> `, so you can`

willnationsdev reviewed (changes requested):
After this round of changes, I think it'll be good for merging.

willnationsdev commented:
I don't have merge rights, so I can't edit the PR directly. :-P

aaronfranke commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -1,127 +1,66 @@
-        {
-            public Node Child { get; set; }
+    Objects check their attached script before the database, which is why scripts can override
+    built-in methods. If a script defines a ``_get_property_list()`` method, Godot merges that data
```
I think explaining how `_get_property_list()` works is too out-of-scope for an article about applying object-oriented principles. It's also something very rarely used in projects, I've never used it, and I just did a recursive grep for all the projects on my computer and the only instance is in a GDNative project.

willnationsdev replied:
Well, the specific subsection within the article is talking about the relationship between a Script and its contributions to the reflection data collected from ClassDB, so it's *kinda* on topic in the sense that it describes further how Scripts contribute to class information.

aaronfranke commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -1,127 +1,66 @@
-start with the ``extends`` keyword, implicitly inherit from the engine's base
-:ref:`Reference <class_Reference>` class. This allows the Object to defer
-to the script's content where the engine logic deems appropriate.
+Applying Object-Oriented principles in Godot
```
Why are the `O` letters capitalized?

NathanLovato commented:
I just addressed all review points.

NathanLovato commented on getting_started/workflow/best_practices/what_are_godot_classes.rst:
```diff
@@ -1,127 +1,59 @@
+.. note::
 
-Script code like this is much slower than engine-side C++ code. Each change
-makes a separate call to the scripting API which leads to many "look-ups" on the
-back-end to find the logic to execute.
+    Even scripts that don't use the ``extends`` keyword implicitly inherit from the engine's base
+    :ref:`Reference <class_Reference>` class. As a result, you can instantiate scripts without the
+    ``extends`` keyword from code. Since they extend ``Reference`` though, you cannot attach them to
+    a :ref:`Node <class_Node>`.
```
@willnationsdev @aaronfranke I'll just ping you to check this note. I trimmed it with the intention of keeping the article as lean and focused on the topic as possible. The fact GDScript files implicitly extend Reference seems important to repeat in some places around the docs, to me, as it can be a source of confusion.

NathanLovato commented:
@aaronfranke thanks for the suggestions, I added them to the PR!

aaronfranke reviewed (approved):
This is a vast improvement over the old article, and I don't have any further suggestions at the moment.

NathanLovato commented:
Thanks much for taking the time to review it!
